net/node.h

```diff
@@ -0,0 +1,104 @@
+/*
+ * An abstract class inherited by Server and Client classes.
+ */
+
+#pragma once
+
+// C++ includes
+#include <map>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+// GeNN robotics includes
+#include "../common/threadable.h"
+
+// local includes
+#include "socket.h"
+
+namespace GeNNRobotics {
+namespace Net {
+class Node; // forward declaration
+
+class Handler
+{
+public:
+    virtual std::string getHandledCommandName() const = 0;
+    virtual void onCommandReceived(Node &node, Command &command) = 0;
+    virtual void onConnected(Node &node) {}
+};
+
+class Node : public Threadable
+{
+public:
+    virtual Socket *getSocket() const = 0;
+
+    /*
+     * Add a handler for a specified type of command (e.g. if it's an IMG command,
+     * it should be handled by Video::NetSource).
+     */
+    void addHandler(Handler &handler)
```

Comment:
The only limitation I see with this is that it only allows a single object to handle one type of event. An alternative would be to just define a handler type as ``typedef std::function<void(Command&)> CommandFn `` (or whatever the syntax is) and **that** is what is passed to ``addHandler(const std::string &command, CommandFn fn)`` rather than a handler object. Then handlers could just call ``addHandler`` to hook their member functions to specific commands. Not 100% sure what you'd do about ``onConnected`` though

Reply:
It's true that this would be a cleaner interface. OTOH all the places where we're handling `Net::Command`s will be encapsulated in classes anyway, so we don't necessarily need the flexibility. But it would be a bit nicer and I probably already have gone overboard with the multiple inheritance in places... We could have a separate set of event handlers for `onConnected` (à la Java).

Reply:
Removing the ``getHandledCommandName()`` method and passing the name to ``addHandler`` would probably also be a simpler way to achieve the same flexibility.